File: sentinelai-frontend/src/types/asset.ts
export interface Asset {
  id: string;
  hostname: string;
  ip_address: string | null;
  mac_address: string | null;
  os: string | null;
  os_version: string | null;
  asset_type: string;
  criticality: string;
  environment: string | null;
  status: string;
  tags: string[] | null;
  vulnerability_count: number;
  open_ports: number;
  last_seen: string | null;
  location: string | null;
  department: string | null;
  owner: string | null;
  vendor: string | null;
  serial_number: string | null;
  notes: string | null;
  risk_score: number;
  discovery_source: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | null;
}

export interface AssetListItem {
  id: string;
  hostname: string;
  ip_address: string | null;
  os: string | null;
  asset_type: string;
  criticality: string;
  status: string;
  risk_score: number;
  department: string | null;
  owner: string | null;
  vulnerability_count: number;
  last_seen: string | null;
  created_at: string;
}

export interface AssetDetail extends Asset {
  risk_details: AssetRiskDetail | null;
  incident_count: number;
  alert_count: number;
  ioc_count: number;
  threat_intel_count: number;
  ai_report_count: number;
  relationships: AssetRelationshipItem[];
  history: AssetHistoryItem[];
}

export interface AssetRiskDetail {
  risk_score: number;
  open_incidents: number;
  critical_alerts: number;
  threat_intel_matches: number;
  cve_count: number;
  exposure_score: number;
  criticality_weight: number;
  calculated_at: string;
}

export interface AssetRelationshipItem {
  id: string;
  source_asset_id: string;
  target_asset_id: string;
  relationship_type: string;
  direction: string;
  target: {
    id: string;
    hostname?: string;
    ip_address?: string;
    asset_type?: string;
    criticality?: string;
  };
}

export interface AssetHistoryItem {
  id: string;
  asset_id: string;
  field_name: string;
  old_value: string | null;
  new_value: string | null;
  changed_by: string | null;
  created_at: string;
}

export interface AssetFilter {
  search?: string;
  asset_type?: string;
  criticality?: string;
  status?: string;
  department?: string;
  owner?: string;
  os?: string;
  risk_level?: string;
  tag?: string;
  sort_by?: string;
  sort_order?: string;
}

export interface AssetStats {
  total_assets: number;
  healthy_assets: number;
  critical_assets: number;
  offline_assets: number;
  high_risk_assets: number;
  by_department: Record<string, number>;
  by_os: Record<string, number>;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  by_criticality: Record<string, number>;
  risk_distribution: Record<string, number>;
}

export interface PaginatedData<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T;
}

File: sentinelai-backend/app/schemas/asset.py
from datetime import datetime
from typing import Any

from pydantic import Field
from app.schemas.base import BaseSchema


class AssetCreate(BaseSchema):
    hostname: str
    ip_address: str | None = None
    mac_address: str | None = None
    os: str | None = None
    os_version: str | None = None
    asset_type: str = "server"
    criticality: str = "medium"
    environment: str | None = None
    status: str = "unknown"
    tags: list[str] | None = None
    location: str | None = None
    department: str | None = None
    owner: str | None = None
    vendor: str | None = None
    serial_number: str | None = None
    notes: str | None = None


class AssetUpdate(BaseSchema):
    hostname: str | None = None
    ip_address: str | None = None
    mac_address: str | None = None
    os: str | None = None
    os_version: str | None = None
    asset_type: str | None = None
    criticality: str | None = None
    environment: str | None = None
    status: str | None = None
    tags: list[str] | None = None
    location: str | None = None
    department: str | None = None
    owner: str | None = None
    vendor: str | None = None
    serial_number: str | None = None
    notes: str | None = None


class AssetResponse(BaseSchema):
    id: str
    hostname: str
    ip_address: str | None = None
    mac_address: str | None = None
    os: str | None = None
    os_version: str | None = None
    asset_type: str
    criticality: str
    environment: str | None = None
    status: str
    tags: list[str] | None = None
    vulnerability_count: int = 0
    open_ports: int = 0
    last_seen: datetime | None = None
    location: str | None = None
    department: str | None = None
    owner: str | None = None
    vendor: str | None = None
    serial_number: str | None = None
    notes: str | None = None
    risk_score: float = 0.0
    discovery_source: str | None = None
    created_at: datetime
    updated_at: datetime
    created_by: str | None = None


class AssetListItem(BaseSchema):
    id: str
    hostname: str
    ip_address: str | None = None
    os: str | None = None
    asset_type: str
    criticality: str
    status: str
    risk_score: float = 0.0
    department: str | None = None
    owner: str | None = None
    vulnerability_count: int = 0
    last_seen: datetime | None = None
    created_at: datetime


class AssetDetailResponse(AssetResponse):
    risk_details: dict[str, Any] | None = None
    incident_count: int = 0
    alert_count: int = 0
    ioc_count: int = 0
    threat_intel_count: int = 0
    ai_report_count: int = 0
    relationships: list[dict[str, Any]] = []
    history: list[dict[str, Any]] = []


class AssetStats(BaseSchema):
    total_assets: int = 0
    healthy_assets: int = 0
    critical_assets: int = 0
    offline_assets: int = 0
    high_risk_assets: int = 0
    by_department: dict[str, int] = {}
    by_os: dict[str, int] = {}
    by_type: dict[str, int] = {}
    by_status: dict[str, int] = {}
    by_criticality: dict[str, int] = {}
    risk_distribution: dict[str, int] = {}


class AssetRelationshipCreate(BaseSchema):
    source_asset_id: str
    target_asset_id: str
    relationship_type: str
    metadata_json: dict[str, Any] | None = None


class AssetImportResult(BaseSchema):
    imported: int = 0
    updated: int = 0
    failed: int = 0
    errors: list[str] = []
